apps/rush-lib/src/logic/buildCache/CacheEntryId.ts

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+const OPTIONS_ARGUMENT_NAME: string = 'options';
+
+export interface IGenerateCacheEntryIdOptions {
+  projectName: string;
+  projectStateHash: string;
+}
+
+export type GetCacheEntryIdFunction = (options: IGenerateCacheEntryIdOptions) => string;
+
+const HASH_TOKEN_NAME: string = 'hash';
+const PROJECT_NAME_TOKEN_NAME: string = 'projectName';
+
+export class CacheEntryId {
+  private constructor() {}
+
+  public static parsePattern(pattern?: string): GetCacheEntryIdFunction {
+    if (!pattern) {
+      return ({ projectStateHash }) => projectStateHash;
+    } else {
```

Comment:
While I want to say that this feels like a lot of effort to parse a few known tokens... The state machine does set it up to be easily expanded in the future. My only other thought would be a very tight regex with capture groups for the `[blah]` tokens, replacing the capture groups, and then validating the re-composed string as a whole as you already do.

Reply:
Yeah that is simpler.